optuna/integration/pytorch_lightning.py

```diff
@@ -56,3 +59,87 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
         if self._trial.should_prune():
             message = "Trial was pruned at epoch {}.".format(epoch)
             raise optuna.TrialPruned(message)
+
+
+class PyTorchLightningDDPPruningCallback(Callback):
+    """PyTorch Lightning DDP callback to prune unpromising trials.
+
+    See `the example <https://github.com/optuna/optuna-examples/blob/
+    main/pytorch/pytorch_lightning_simple.py>`__
+    if you want to add a pruning callback which observes accuracy.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        monitor:
+            An evaluation metric for pruning, e.g., ``val_loss`` or
+            ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
+            ``pytorch_lightning.LightningModule.training_step`` or
+            ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
+            how this dictionary is formatted.
+    """
+
+    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
+        _imports.check()
+
+        if not isinstance(trial.study._storage, _CachedStorage):
+            raise ValueError(
+                ":class:`~optuna.integration.PyTorchLightningPruningCallback`"
+                " supports only :class:`~optuna.storages.RDBStorage`."
+            )
+
+        super().__init__()
+
+        self._trial = trial
+        self.monitor = monitor
+
+    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        epoch = pl_module.current_epoch
+
+        current_score = trainer.callback_metrics.get(self.monitor)
+        if current_score is None:
+            message = (
+                "The metric '{}' is not in the evaluation logs for pruning. "
+                "Please make sure you set the correct metric name.".format(self.monitor)
+            )
+            warnings.warn(message)
+            return
+
+        # When multi gpu, on_validation_end function executes multi times.
+        # So Exception is ignored to avoid IntegrityError.
+        distributed_backend = trainer.accelerator_connector.distributed_backend
+        try:
+            self._trial.report(current_score, step=epoch)
+        except sqlalchemy.exc.IntegrityError:
+            if distributed_backend is not None:
+                pass
+            else:
+                raise
```

Comment:
How about this?

```suggestion
        if trainer.is_global_zero:
            self._trial.report(current_score, step=epoch)
```
```

Reply:
I should have explained a little bit more. In my understanding, the error case is caused when the multiple processes try to call the `report` function at each `epoch`. We can handle this case by calling the function `global_zero`'s process because the global rank zero is a unique process among distributed data-parallel training.